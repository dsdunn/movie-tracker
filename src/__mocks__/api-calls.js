export const signUp = jest.fn().mockImplementation((state) => {
  if(state.name === 'bugs') {
    return Promise.resolve({id: 1})
  } else {
    return 'nothing'
  }
}); 
