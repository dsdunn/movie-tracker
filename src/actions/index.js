export const addMovies = (movies) => ({
  type: "ADD_MOVIES",
  movies
});

export const toggleFav = (id) => {
  type: "TOGGLE_FAV",
  id
};

export const toggleFilter = () => {
  type: "TOGGLE_FILTER"
};