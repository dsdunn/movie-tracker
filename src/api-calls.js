import { apiKey } from './apiKey';

export const getMovies = async () => {
  try {
    const url = `https://api.themoviedb.org/3/discover/movie?api_key=${apiKey}&language=en-US&sort_by=popularity.desc&include_adult=false&include_video=false&page=1&primary_release_year=2018`;
    const response = await fetch(url)
    return response.json();
  } 
  catch(error) {
    return error.message;
  }
}

export const logIn = async (state) => {
  try{
    const url = 'http://localhost:3000/api/users/';
    const {email, password} = state;

    const response = await fetch(url, {
      method: 'POST',
      body: JSON.stringify({email, password}),
      headers:{
        'Content-Type': 'application/json'
      }
    })
    const result = await response.json();
    return result.data;
 
  } 
  catch (error) {
    return false;
  }
}

export const signUp = async (state) => {
  try{
    const url = 'http://localhost:3000/api/users/new';
    const {name, email, password} = state;

    const response = await fetch(url, {
      method: 'POST',
      body: JSON.stringify({name, email, password}),
      headers:{
        'Content-Type': 'application/json'
      }
    });
    return await response.json();
  } 
  catch (error) {
    return false;
  }
}

export const postFavorite = async (favorite) => {
  console.log(favorite)
  try {
    const url = 'http://localhost:3000/api/users/favorites/new';
    const response = await fetch(url, {
      method: 'POST',
      body: JSON.stringify(favorite),
      headers: {
        'Content-Type': 'application/json'
      } 
    });
    return await response.json();
  }
  catch (error) {
    return false;
  }
}
















