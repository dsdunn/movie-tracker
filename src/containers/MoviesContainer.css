.movie-container {
  display: flex;
  overflow: auto;
  white-space: nowrap;
}

.movie-container div {
  padding: 1em;
  background-color: #607D8B;
  color: #212121;
  border: 2px solid #212121;
  border-radius: 20px;
  margin: 1em;
}

.movie-container div button {
  background-color: #009688;
  border: 0;
  border-radius: 20px;
  font-size: 1em;
  color: #FFFFFF;
  padding: .5em 1em;
}