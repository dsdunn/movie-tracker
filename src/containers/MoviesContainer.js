import React from 'react';
import { connect } from 'react-redux';
import { MovieCard } from '../components/MovieCard.js';
import { postFavorite, deleteDatabaseFav } from '../api-calls.js';
import { addLocalFav, deleteLocalFav } from '../actions';
import PropTypes from 'prop-types';
import './MoviesContainer.css';

export const MoviesContainer = (props) => {

  const toggleFav = async (userId, movie) => {
    if(!props.user.name) {
      props.history.push('/login');
      return null;
    }
    if (isDuplicate(movie.movie_id)) {  
      props.deleteLocalFav(movie.movie_id);
      await deleteDatabaseFav(userId, movie.movie_id);
    } else { 
      const newFavorite = {user_id: userId, ...movie};
      await postFavorite(newFavorite);
      props.addLocalFav(movie.movie_id);
    }
  };


  const isDuplicate = (movieId) => {
    return props.user.favorites.find(favoriteId => favoriteId === movieId);
  };

  const isFavorite = (id) => {
    return props.user.favorites ? 
    props.user.favorites.includes(id) : null;
  }
  
  const cards = props.movies.map(movie => (
    <MovieCard
      movie={movie} 
      key={movie.id} 
      userId={props.user.id}
      toggleFav={toggleFav}
      isFav={isFavorite(movie.movie_id)}
    />)
  );

  const favoriteCards = props.movies.map(movie => {
    if (isFavorite(movie.movie_id)) {
      return ( 
        <MovieCard
          movie={movie} 
          key={movie.id} 
          userId={props.user.id}
          toggleFav={toggleFav}
          isFav={true}
        />
      )
    }
  })

  return (
    <div className='movie-container'>
      {props.showAllMovies  ? cards : favoriteCards}
    </div>
  );
};

export const mapStateToProps = (state) => ({
  movies: state.movies,
  showAllMovies: state.showAllMovies,
  user: state.user
});

export const mapDispatchToProps = (dispatch) => ({
  addLocalFav: (movie) => dispatch(addLocalFav(movie)),
  deleteLocalFav: (movieId) => dispatch(deleteLocalFav(movieId))  
});

MoviesContainer.propTypes = {
  addLocalFav: PropTypes.func,
  deleteLocalFav: PropTypes.func,
  user: PropTypes.object,
  movies: PropTypes.array
};

export default connect(mapStateToProps, mapDispatchToProps)(MoviesContainer);