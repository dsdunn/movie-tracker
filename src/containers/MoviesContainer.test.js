import React from 'react';
import { MoviesContainer, mapStateToProps, mapDispatchToProps } from './MoviesContainer.js'
import * as actions from '../actions'
import { shallow, mount } from 'enzyme';
import {deleteDatabaseFav, postFavorite} from '../api-calls';
jest.mock('../api-calls');



describe('MoviesContainer', () => {
  let wrapper, 
  mockDeleteLocalFav, 
  mockAddLocalFav, 
  mockMovies,
  mockShowAllMovies,
  mockHistory,
  mockUser;

  beforeEach(() => {
    mockDeleteLocalFav = jest.fn();
    mockAddLocalFav = jest.fn();
    mockUser = {name:'oscar', id: 4, favorites: [1, 4]}
    mockMovies = [{movie_id: 1}]
    mockShowAllMovies = true;
    mockHistory = []

    wrapper = mount(<MoviesContainer 
      deleteLocalFav={mockDeleteLocalFav}
      addLocalFav={mockAddLocalFav}
      showAllMovies={mockShowAllMovies}
      user={mockUser}
      movies={mockMovies}
      history={mockHistory}
      />)

  }) 

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('when toggleFav is called, if the movie id already exists in state, deleteLocalFav should be called', () => {
    wrapper.find('button').simulate('click')

    expect(mockDeleteLocalFav).toBeCalled()
  })

  it('when toggleFave is called it should call deleteDatabaseFav if movie_id already exists in state', async() => {
    await wrapper.find('button').simulate('click')
    expect(deleteDatabaseFav).toBeCalled();
  })

  it('should push login to history when there is no user logged in and toggleFav is called', () => {
    wrapper = mount(
      <MoviesContainer 
      deleteLocalFav={mockDeleteLocalFav}
      addLocalFav={mockAddLocalFav}
      showAllMovies={mockShowAllMovies}
      user={false}
      movies={mockMovies}
      history={mockHistory}
      />)
    wrapper.find('button').simulate('click')
    expect(mockHistory).toEqual(['/login'])

  })

  it('should add a new favorite to state if the movie_id doesnt already exist', () => {
    mockMovies = [{movie_id: 2}]
    wrapper = mount(<MoviesContainer 
      deleteLocalFav={mockDeleteLocalFav}
      addLocalFav={mockAddLocalFav}
      showAllMovies={mockShowAllMovies}
      user={mockUser}
      movies={mockMovies}
      history={[]}
    />)
    wrapper.find('button').simulate('click')
    expect(postFavorite).toBeCalled();
  })
   
})
