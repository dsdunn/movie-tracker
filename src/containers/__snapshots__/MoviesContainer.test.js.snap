// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviesContainer should match the snapshot 1`] = `
ReactWrapper {
  "length": 1,
  Symbol(enzyme.__unrendered__): <Unknown
    addLocalFav={[Function]}
    deleteLocalFav={[Function]}
    history={Array []}
    movies={
        Array [
            Object {
              "movie_id": 1,
            },
          ]
    }
    showAllMovies={true}
    user={
        Object {
            "favorites": Array [
              1,
              4,
            ],
            "id": 4,
            "name": "oscar",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "addLocalFav": [Function],
      "deleteLocalFav": [Function],
      "history": Array [],
      "movies": Array [
        Object {
          "movie_id": 1,
        },
      ],
      "showAllMovies": true,
      "user": Object {
        "favorites": Array [
          1,
          4,
        ],
        "id": 4,
        "name": "oscar",
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": <div
        class="movie-container"
>
        <div
                class="movie-card favorite"
        >
                <img
                        alt="popular movie"
                        src="https://image.tmdb.org/t/p/w300/undefined"
                />
                <h3 />
                <h3 />
                <button>
                        Favorite
                </button>
        </div>
</div>,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Unknown
            isFav={true}
            movie={
                        Object {
                                    "movie_id": 1,
                                  }
            }
            toggleFav={[Function]}
            userId={4}
/>,
        ],
        "className": "movie-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "isFav": true,
            "movie": Object {
              "movie_id": 1,
            },
            "toggleFav": [Function],
            "userId": 4,
          },
          "ref": null,
          "rendered": Object {
            "instance": <div
              class="movie-card favorite"
>
              <img
                            alt="popular movie"
                            src="https://image.tmdb.org/t/p/w300/undefined"
              />
              <h3 />
              <h3 />
              <button>
                            Favorite
              </button>
</div>,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <img
                  alt="popular movie"
                  src="https://image.tmdb.org/t/p/w300/undefined"
/>,
                <h3 />,
                <h3 />,
                <button
                  onClick={[Function]}
>
                  Favorite
</button>,
              ],
              "className": "movie-card favorite",
              "id": undefined,
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": <img
                  alt="popular movie"
                  src="https://image.tmdb.org/t/p/w300/undefined"
/>,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "alt": "popular movie",
                  "src": "https://image.tmdb.org/t/p/w300/undefined",
                },
                "ref": null,
                "rendered": Array [
                  undefined,
                ],
                "type": "img",
              },
              Object {
                "instance": <h3 />,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": undefined,
                },
                "ref": null,
                "rendered": Array [
                  undefined,
                ],
                "type": "h3",
              },
              Object {
                "instance": <h3 />,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": undefined,
                },
                "ref": null,
                "rendered": Array [
                  undefined,
                ],
                "type": "h3",
              },
              Object {
                "instance": <button>
                  Favorite
</button>,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Favorite",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": Array [
                  "Favorite",
                ],
                "type": "button",
              },
            ],
            "type": "div",
          },
          "type": [Function],
        },
      ],
      "type": "div",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "addLocalFav": [Function],
        "deleteLocalFav": [Function],
        "history": Array [],
        "movies": Array [
          Object {
            "movie_id": 1,
          },
        ],
        "showAllMovies": true,
        "user": Object {
          "favorites": Array [
            1,
            4,
          ],
          "id": 4,
          "name": "oscar",
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": <div
          class="movie-container"
>
          <div
                    class="movie-card favorite"
          >
                    <img
                              alt="popular movie"
                              src="https://image.tmdb.org/t/p/w300/undefined"
                    />
                    <h3 />
                    <h3 />
                    <button>
                              Favorite
                    </button>
          </div>
</div>,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Unknown
              isFav={true}
              movie={
                            Object {
                                          "movie_id": 1,
                                        }
              }
              toggleFav={[Function]}
              userId={4}
/>,
          ],
          "className": "movie-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "isFav": true,
              "movie": Object {
                "movie_id": 1,
              },
              "toggleFav": [Function],
              "userId": 4,
            },
            "ref": null,
            "rendered": Object {
              "instance": <div
                class="movie-card favorite"
>
                <img
                                alt="popular movie"
                                src="https://image.tmdb.org/t/p/w300/undefined"
                />
                <h3 />
                <h3 />
                <button>
                                Favorite
                </button>
</div>,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <img
                    alt="popular movie"
                    src="https://image.tmdb.org/t/p/w300/undefined"
/>,
                  <h3 />,
                  <h3 />,
                  <button
                    onClick={[Function]}
>
                    Favorite
</button>,
                ],
                "className": "movie-card favorite",
                "id": undefined,
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": <img
                    alt="popular movie"
                    src="https://image.tmdb.org/t/p/w300/undefined"
/>,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "alt": "popular movie",
                    "src": "https://image.tmdb.org/t/p/w300/undefined",
                  },
                  "ref": null,
                  "rendered": Array [
                    undefined,
                  ],
                  "type": "img",
                },
                Object {
                  "instance": <h3 />,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": undefined,
                  },
                  "ref": null,
                  "rendered": Array [
                    undefined,
                  ],
                  "type": "h3",
                },
                Object {
                  "instance": <h3 />,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": undefined,
                  },
                  "ref": null,
                  "rendered": Array [
                    undefined,
                  ],
                  "type": "h3",
                },
                Object {
                  "instance": <button>
                    Favorite
</button>,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Favorite",
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": Array [
                    "Favorite",
                  ],
                  "type": "button",
                },
              ],
              "type": "div",
            },
            "type": [Function],
          },
        ],
        "type": "div",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
